### ISSUE TITLE
Handling of compatible structure types under C23 semantics

### ISSUE THREAD

**Issue by jprotopopov-ut** [goblint/analyzer#1950]
C23 has relaxed rules for `struct` type compatibility making the following code snippet legal:
```c
typedef void (*func_t)(void);

struct A {
    func_t f;
};

void test1(struct A {
    func_t f;
} *b) {
    b->f();
}

void dummy() {}

int main() {
    struct A arg = {dummy};
    test1(&arg);
    return 0;
}
```
Goblint (revision e816b4aa7baea44ce2e973b58e39145743eea377) fails to recognize that `b->f` is initialized to `dummy` and outputs
```
[Warning][Unsound][Call] No suitable function to be called at call site. Continuing with state before call
```

[C23 standard](https://open-std.org/JTC1/SC22/WG14/www/docs/n3220.pdf) (section 6.2.7) specifies:
```
Moreover, two complete structure, union, or enumerated types declared with the
same tag are compatible if members satisfy the following requirements:
— there shall be a one-to-one correspondence between their members such that each pair of
corresponding members are declared with compatible types;
— if one member of the pair is declared with an alignment specifier, the other is declared with an
equivalent alignment specifier;
— and, if one member of the pair is declared with a name, the other is declared with the same
name
```
More discussion is also available in the following [proposal](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3037.pdf).

This issue has surfaced when trying to analyze [rsync](https://gitlab.com/analyzethat/2026/-/raw/main/rsync-v3.4.1.i?ref_type=heads) version from AnalyzeThat. The linked code is in fact even more liberal in using pointers to incompatible struct types, as it declares these struct types with different tags (for example, see the type of the first parameter in the call to `deflateInit_` from `compress2`).
